sycl/include/sycl/detail/vector_arith.hpp

```diff
@@ -24,49 +24,166 @@ template <typename DataT, int NumElem> class __SYCL_EBO vec;
+          // "convert" back to 0/1.
+          //
+          // Some operations are known to produce only 0/1 for valid bool inputs
+          // though, no reason to do extra processing for them:
+          if constexpr (!check_type_in_v<BinOp, std::multiplies<void>,
+                                         std::divides<void>, std::bit_or<void>,
+                                         std::bit_and<void>, std::bit_xor<void>,
+                                         ShiftRight>) {
+            for (size_t i = 0; i < N; ++i)
+              res[i] = bit_cast<int8_t>(res[i]) != 0;
+          }
+        }
+        return result_t{(typename result_t::vector_t)res};
+      }
+    }
   }
-#else // __SYCL_DEVICE_ONLY__
 
-#define BINOP_BASE(BINOP, OPASSIGN, CONVERT, COND)                             \
+  template <typename UnaryOp, typename Self>
+  static constexpr auto apply_unary_op(const Self &self) {
+    static_assert(is_vec_v<Self>);
+    constexpr bool is_logical = std::is_same_v<UnaryOp, std::logical_not<void>>;
+    using element_type = typename Self::element_type;
+    constexpr int N = Self::size();
+    using result_t = std::conditional_t<
+        is_logical, vec<fixed_width_signed<sizeof(element_type)>, N>, Self>;
+    UnaryOp Op{};
+    if constexpr (is_host ||
+                  std::is_same_v<element_type, ext::oneapi::bfloat16>) {
+      result_t res{};
+      for (size_t i = 0; i < N; ++i)
+        if constexpr (is_logical)
+          res[i] = Op(self[i]) ? -1 : 0;
+        else
+          res[i] = Op(self[i]);
+      return res;
+    } else {
+      using vector_t = typename Self::vector_t;
+      auto res = Op(bit_cast<vector_t>(self));
+      if constexpr (is_logical) {
+        if constexpr (N == 1) {
+          // vector_t for one-element vector is just scalar with normal C++
+          // semantics, need to align it with vector behavior:
+          return result_t{(typename result_t::vector_t)res * -1};
+        }
+        return result_t{(typename result_t::vector_t)res};
+      } else {
+        if constexpr (std::is_same_v<element_type, bool>) {
+          // OpenCL/SPIR-V has different semantics for operations on bool (which
+          // is mapped to some 8-bit integer type on device) than SYCL. Need to
+          // "convert" back to 0/1.
+          //
+          // Some operations are known to produce only 0/1 for valid bool inputs
+          // though, no reason to do extra processing for them:
+          if constexpr (!check_type_in_v<UnaryOp, UnaryPlus>) {
+            for (size_t i = 0; i < N; ++i)
+              res[i] = bit_cast<int8_t>(res[i]) != 0;
+          }
+        }
+        return result_t{(typename result_t::vector_t)res};
+      }
+    }
+  }
+};
+
+// Macros to populate binary operation on sycl::vec.
+#if defined(__SYCL_BINOP)
+#error "Undefine __SYCL_BINOP macro"
+#endif
+
+#define __SYCL_BINOP(BINOP, OPASSIGN, CONVERT, COND, FUNCTOR)                  \
```

Comment:
The `CONVERT` argument of the macro is now unused. I think we can safely remove it.

Reply:
Will do in the next iteration.